Fix MCCAS after changed made to MC layer upstream

rastogishubham opened the pull request:
The MC layer has gone through many changes, which affect MCCAS, especially the changes made to MCFragments. 

This PR addresses them all

rdar://156159206

rastogishubham committed:
Revert "[MCCAS] Add configuration LLVM_ENABLE_MCCAS"

This reverts commit 1d94ada9141d7c162eb4327daa12fd33dbe33080.

rastogishubham committed:
Remove MCPseudoProbeAddrFragment from MCCASObjectV1.def

The MCPseudoProbeAddrFragment type has been removed as a type of
fragment and has been merged with the MCLEBFragment.

rastogishubham committed:
Remove MCSectionCAS class functions that no longer exist in the base class

The MCSection class has been refactored to no longer have
printSwitchToSection, and useCodeAlign function. It also no longer can
contain a classof function. This change removes these functions to make
it compile properly again.

rastogishubham committed:
Change MCFRAGMENT_NODE_REF to use MCFragment

The create function in the MCFRAGMENT_NODE_REF macro has to take an
MCFragment object because of the changes in MCFragment which removed
the concept that different fragment types would have their own class.

rastogishubham committed:
Fix the MCAlignFramgnetRef class.

The MCAlignFragment class has been removed from MCSection.h and the
Align Fragment is now part of the MCFragment class.

The Align fragment has also been significantly changed. In the past, the
Align fragment would only contain alignment data, however, now it can
also contain regular data as well (similar to an MCDataFragment, but
with alignment data as well). This patch makes the necessary changes to
make MCCAS work properly with the new MCAlignFragment.

rastogishubham committed:
Fix MCBoundaryAlignFragmentRef::create.

Since the create function in MCFRAGMENT_NODE_REF just takes an
MCFragment object as an argument, this change just makes the change for
MCBoundaryAlignFragmentRef::create.

rastogishubham committed:
Fix MCCVInlineLineTableFragmentRef::create.

Since the create function in MCFRAGMENT_NODE_REF just takes an
MCFragment object as an argument, this change just makes the change for
MCCVInlineLineTableFragmentRef::create.

rastogishubham committed:
Fix MCFillFragmentRef::create.

Since the create function in MCFRAGMENT_NODE_REF just takes an
MCFragment object as an argument, this change just makes the necessary
changes for MCFillFragmentRef::create.

rastogishubham committed:
Fix MCLEBFragmentRef::create.

Since the create function in MCFRAGMENT_NODE_REF just takes an
MCFragment object as an argument, this change just makes the change for
MCLEBFragmentRef::create.

rastogishubham committed:
Fix MCNopsFragmentRef::create.

Since the create function in MCFRAGMENT_NODE_REF just takes an
MCFragment object as an argument, this change just makes the necessary
changes for MCNopsFragmentRef::create.

rastogishubham committed:
Fix MCOrgFragmentRef::create.

Since the create function in MCFRAGMENT_NODE_REF just takes an
MCFragment object as an argument, this change just makes the necessary
changes for MCOrgFragmentRef::create.

rastogishubham committed:
Fix MCSymbolIdFragmentRef::create.

Since the create function in MCFRAGMENT_NODE_REF just takes an
MCFragment object as an argument, this change just makes the necessary
changes for MCSymbolIdFragmentRef::create.

rastogishubham committed:
Fix getFragmentContents() function.

The MCFragments have been changed, they can now have both a fixed
content and a variable content as well. This patch makes the necessary
changes to the getFragmentContents() function to make it work properly.

rastogishubham committed:
Fix MCCASBuilder::mergeMCFragmentContents

MCCASBuilder::mergeMCFragmentContents merges all content from all
the fragments in a section into one buffer. However, since the way
content is stored in a fragemnts has changed, this function had to be
updated as well.

rastogishubham committed:
Update bool IsMergeableFragment.

There is no concept on an EncodedFragment anymore, hence we have to list
all fragments that are mergeable.

rastogishubham committed:
Remove cast to MCDataFragment.

There is no class called MCDataFragment, it has been folded into the
MCFragment class. Change the cast<MCDataFragment> to cast<MCFragment>.

rastogishubham committed:
Change isVirtualSection to isBssSection

rastogishubham committed:
Fix any broken tests

rastogishubham committed:
Added TotalFragmentWithoutAddendsSize to make debugging easier.

rastogishubham commented on llvm/lib/MCCAS/MCCASObjectV1.cpp:
```diff
@@ -3123,6 +3123,13 @@ Error MCCASBuilder::buildFragments() {
+    // CASObjects, one can use a conditional breakpoint to find the Fragment
+    // which might have a bug.
+#ifndef NDEBUG
+    uint64_t TotalFragmentWithoutAddendsSize = 0;
```
I do get the warning 

```
[236/236] Building CXX object lib/MCCAS/CMakeFiles/LLVMMCCAS.dir/MCCASObjectV1.cpp.o
/Users/shubhamrastogi/Development/llvm-project-cas/llvm-project/llvm/lib/MCCAS/MCCASObjectV1.cpp:3131:14: warning: variable 'TotalFragmentWithoutAddendsSize' set but not used [-Wunused-but-set-variable]
 3131 |     uint64_t TotalFragmentWithoutAddendsSize = 0;
      |              
```

This is because `TotalFragmentWithoutAddendsSize` is never read from, but this is really useful for debugging MCCAS. Any ideas on what to do?

cachemeifyoucan replied:
I suggest use debug macros: https://llvm.org/docs/ProgrammersManual.html#the-llvm-debug-macro-and-debug-option

For non-assert build just add a reference:
```
(void) TotalFragmentWithoutAddendsSize;
```

rastogishubham replied:
Does it look okay now?

cachemeifyoucan replied:
You want to add `#define DEBUG_TYPE` for the file.

rastogishubham replied:
I already did 
https://github.com/swiftlang/llvm-project/pull/11184/commits/8d6a93d30978394aa2d486950ac0874b616bca56#diff-a3c1d278b91afb9b05b269ccd8a53cdfa818c281edd70e3b7407590dd910199aR44

adrian-prantl commented on llvm/lib/MCCAS/MCCASObjectV1.cpp:
```diff
@@ -1532,43 +1532,58 @@ Expected<uint64_t> AtomRef::materialize(MCCASReader &Reader,
+
+  *Stream << Remaining.substr(0, FragContentSize);
+
+  Remaining = Remaining.drop_front(FragContentSize);
```
If Remaining is a `StringRef`, maybe `consume_front` is a better API here, to both extract and shrink in one go.

rastogishubham replied:
I don't think that is accurate:

```
    /// Returns true if this StringRef has the given prefix and removes that
    /// prefix.
    bool consume_front(StringRef Prefix) {
    
```

Consume only returns true and I would still have to extract the `Prefix` my using `substr`

cachemeifyoucan replied:
I think `drop_front` is correct here.

rastogishubham committed:
Remove DebugStringOffsets section from InMemoryCASDWARFObject.

In rare cases (like in the case of armv7-cas.ll), where we still produce
DWARF4, we are still trying to read the information from a non-existent
DebugStringOffsets section.

We do not need the data from that section to partition compile units.

That code has been removed

cachemeifyoucan commented on llvm/lib/MCCAS/MCCASObjectV1.cpp:
```diff
@@ -1532,43 +1532,58 @@ Expected<uint64_t> AtomRef::materialize(MCCASReader &Reader,
+  uint64_t Count = (FragmentSize - F.getFixedSize()) / F.getAlignFillLen();
+  if (F.hasAlignEmitNops()) {
+    // Write 1 to signify that it has nops.
+    writeVBR8(1, B->Data);
```
You don't need VBR8 to store 1 bit of data. The old serialization is reusing the size field to indicating using no-op but now you can't reuse that anymore.

cachemeifyoucan commented on llvm/lib/MCCAS/MCCASObjectV1.cpp:
```diff
@@ -1532,43 +1532,58 @@ Expected<uint64_t> AtomRef::materialize(MCCASReader &Reader,
     B->Data.append(MB.FragmentData);
     return get(B->build());
   }
+  // Write 0 to signify that it has nops.
```
This comment should be the other way around?

rastogishubham replied:
yeah sorry, I messed up there, good catch!

cachemeifyoucan commented on llvm/lib/MCCAS/MCCASObjectV1.cpp:
```diff
@@ -1653,15 +1653,16 @@ MCCVInlineLineTableFragmentRef::materialize(MCCASReader &Reader,
+MCFillFragmentRef::create(MCCASBuilder &MB, const MCFragment &F,
                           unsigned FragmentSize,
                           ArrayRef<char> FragmentContents) {
+  auto *FillFrag = dyn_cast<MCFillFragment>(&F);
```
This should be a `cast` since if the type is wrong, this will crash anyway.

cachemeifyoucan commented on llvm/lib/MCCAS/MCCASObjectV1.cpp:
```diff
@@ -1721,14 +1721,15 @@ Expected<uint64_t> MCLEBFragmentRef::materialize(MCCASReader &Reader,
+MCNopsFragmentRef::create(MCCASBuilder &MB, const MCFragment &F,
                           unsigned FragmentSize,
                           ArrayRef<char> FragmentContents) {
+  auto *NopsFrag = dyn_cast<MCNopsFragment>(&F);
```
Same here.

cachemeifyoucan commented on llvm/lib/MCCAS/MCCASObjectV1.cpp:
```diff
@@ -1849,8 +1863,7 @@ Error MCCASBuilder::buildFragment(const MCFragment &F, unsigned Size,
   case MCFragment::MCEnumName: {                                               \
-    const MCFragmentName &SF = cast<MCFragmentName>(F);                        \
-    auto FN = MCFragmentName##Ref::create(*this, SF, Size, FragmentContents);  \
+    auto FN = MCFragmentName##Ref::create(*this, F, Size, FragmentContents);   \
```
I think the `cast` should still be here instead of doing `dyn_cast` inside `create` function.

rastogishubham replied:
The reason I can't do that is because some fragments like Align Fragments are no longer their own class, they are just part of MCFragment, so if we are going to use the macro, the create function needs to take an MCFragment, rather than a specific type of fragment

rastogishubham committed:
MCNopsFragment change dyn_cast to cast

rastogishubham committed:
Add test to make sure LEB fragments are handled correctly.
